### ISSUE TITLE
[re-opened] Identify possible NPE for PHP

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#460]
It should identify that your variable could be `null`, then you should do a double check before use that like:

```php
$user = getUser(); // @return User|null
$user->getName(); // Ops!
```

The fix should be:

```php
if ($user !== null) {
    ...
}
```

Or:

```php
/** @var User $user */
$user = getUser();
```

I think that we have something like that implemented, but I can't found. Maybe we just need to improve it.

**Comment by kalessil** [comment]
Oh, that's a nice further extension, thanks!

**Comment by rentalhost** [comment]
Wow! It was fast. I will try it on my codebase at night - about 8h hours from now.

**Comment by rentalhost** [comment]
I am trying the last version, but it seems that do not catch some cases. For instance:

```php
class MyClass
{
    public function get(): ?MyClass
    {
        return null;
    }
}

$myClass = (new MyClass)->get();
$myClass->get();

(new MyClass)->get()->get();
```

The line `$myClass->get()` and `(new MyClass)->get()->get()` should report it, once that `get()` returns a nullable type.

**Comment by kalessil** [comment]
Weird, I'll investigate. Thanks for checking!
Re-opening - need to evaluate the case/extends test cases.

**Comment by rentalhost** [comment]
Oh... why not functions too?

And it not works for the example below:

```php
class MyClass
{
    /** @var self|null */
    public $from;

    public function example(): void
    {
        $couldBeNull = $this->from;
        $couldBeNull->get(); // <-- $this->from is nullable, but nothing is warned here
    }

    public function get(): ?MyClass
    {
        return null;
    }
}
```

**Comment by kalessil** [comment]
Reassigning from properties is also ignored, place the original code in the `example` =)
Do we need functions too, I can check if this feasible?

**Comment by rentalhost** [comment]
I think that is important check variables like this too. Functions are not much common, except for helpers, but you should consider anonymous functions (I have no tried it).

**Comment by kalessil** [comment]
Agree, we'll add fields analysis later, currently, I have no ideas how to identify null-checks for properties.
Functions: as discussed, I'll check how this affects performance - I just finished quartal performance battle...

**Comment by rentalhost** [comment]
What is your difficulties for properties?

**Comment by rentalhost** [comment]
Hey, do you use Telegram? Maybe could help if us (and another developers from team) stay on a chat. It should works better than here. We can create a group for that. If you agree, you can find me `@DavidRodriguesR`.

**Comment by kalessil** [comment]
My response rate is not better with messengers xD + tickets threads are very efficient as holding context.

The with properties is that its; nullability can be checked not in the method which we are inspecting, looking up the whole class costs too much CPU time - that;s it =)

**Comment by rentalhost** [comment]
The `$coukdBeNull` variable don't hold the types already returned from property? Maybe you just need check that.
